plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -219,21 +220,28 @@ ty ~> vTy
 -}
 inferTypeM (Let ann r@NonRec bs inTerm) = do
     -- Check each binding individually, then if ok, introduce its new type/vars to the (linearly) next let or inTerm
-    ty <- foldr checkBindingThenScope (inferTypeM inTerm) bs
+    ty <- foldr checkBindingThenScope (inferTypeM (substTyBinds inTerm)) bs
```

Comment:
I don't think this is quite right. This adds all the bindings into the type environment, and then typechecks the body of the let... but only after _already_ doing the type substitutions! What's the point of checking the body with the type bindings in the environment if we're already substituting up-front?

Importantly, this can make typechecking much more expensive: because we will normalize each substituted type in place, instead of just referring to a variable in the environment.

I think we just want to substitute the type bindings in `ty` after we've inferred it, but before we do the check on line 225.

Reply:
TBH, this is a bit of a dark corner and could do with a Note! In fact, there already is a bit of a note: `PIR vs Paper Escaping Types Differences`. Can you add a reference to it and also add something about how we are going to handle type-lets?

Reply:
> I think we just want to substitute the type bindings in ty after we've inferred it, but before we do the check on line 225.

@michaelpj this is what I tried at first, but it doesn't pass the `typeLet2` test: 

```
(let
  (nonrec)
  (typebind (tyvardecl a (type)) (con integer))
  [ (lam x a x) (con integer 1) ]
)
```

The type checker cannot infer the type of `[ (lam x a x) (con integer 1) ]`, because "expecting `a`, got `(con integer)`".

If I remove the type bindings from `bs` on this line, does that address your concern?

Reply:
That test *should* continue failing! We don't want fully transparent lets, we want them to behave like type abstractions. Type abstractions are also opaque on the inside:
```
(/\a . (\x : a . x) 1) {integer}
```
won't typecheck!

Reply:
Ah, OK